In [1]:
from sklearn.datasets import load_boston
import pandas as pd
import numpy as np

In [2]:
RANDOM_STATE = 42

In [3]:
dataset = load_boston()
X = pd.DataFrame(dataset.data)
X.columns = dataset.feature_names
y = dataset.target

In [4]:
X

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0    0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0   
1    0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0   
2    0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  242.0   
3    0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622  3.0  222.0   
4    0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622  3.0  222.0   
..       ...   ...    ...   ...    ...    ...   ...     ...  ...    ...   
501  0.06263   0.0  11.93   0.0  0.573  6.593  69.1  2.4786  1.0  273.0   
502  0.04527   0.0  11.93   0.0  0.573  6.120  76.7  2.2875  1.0  273.0   
503  0.06076   0.0  11.93   0.0  0.573  6.976  91.0  2.1675  1.0  273.0   
504  0.10959   0.0  11.93   0.0  0.573  6.794  89.3  2.3889  1.0  273.0   
505  0.04741   0.0  11.93   0.0  0.573  6.030  80.8  2.5050  1.0  273.0   

     PTRATIO       B  LSTAT  
0       15.3  396.90   4.98  
1       17.8  396.90   9.14  
2       17.8  392.83   4.03  
3       18.7  394.63   2.94  
4       18.7  396.90   5.33  
..       ...     ...    ...  
501     21.0  391.99   9.67  
502     21.0  396.90   9.08  
503     21.0  396.90   5.64  
504     21.0  393.45   6.48  
505     21.0  396.90   7.88  

[506 rows x 13 columns]

In [5]:
y

array([24. , 21.6, 34.7, 33.4, 36.2, 28.7, 22.9, 27.1, 16.5, 18.9, 15. ,
       18.9, 21.7, 20.4, 18.2, 19.9, 23.1, 17.5, 20.2, 18.2, 13.6, 19.6,
       15.2, 14.5, 15.6, 13.9, 16.6, 14.8, 18.4, 21. , 12.7, 14.5, 13.2,
       13.1, 13.5, 18.9, 20. , 21. , 24.7, 30.8, 34.9, 26.6, 25.3, 24.7,
       21.2, 19.3, 20. , 16.6, 14.4, 19.4, 19.7, 20.5, 25. , 23.4, 18.9,
       35.4, 24.7, 31.6, 23.3, 19.6, 18.7, 16. , 22.2, 25. , 33. , 23.5,
       19.4, 22. , 17.4, 20.9, 24.2, 21.7, 22.8, 23.4, 24.1, 21.4, 20. ,
       20.8, 21.2, 20.3, 28. , 23.9, 24.8, 22.9, 23.9, 26.6, 22.5, 22.2,
       23.6, 28.7, 22.6, 22. , 22.9, 25. , 20.6, 28.4, 21.4, 38.7, 43.8,
       33.2, 27.5, 26.5, 18.6, 19.3, 20.1, 19.5, 19.5, 20.4, 19.8, 19.4,
       21.7, 22.8, 18.8, 18.7, 18.5, 18.3, 21.2, 19.2, 20.4, 19.3, 22. ,
       20.3, 20.5, 17.3, 18.8, 21.4, 15.7, 16.2, 18. , 14.3, 19.2, 19.6,
       23. , 18.4, 15.6, 18.1, 17.4, 17.1, 13.3, 17.8, 14. , 14.4, 13.4,
       15.6, 11.8, 13.8, 15.6, 14.6, 17.8, 15.4, 21

In [6]:
from sklearn.model_selection import train_test_split

1. Разделите выборку на обучающую и тестовую в отношении 80%/20%

In [7]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=RANDOM_STATE)

In [8]:
X_train.shape, X_test.shape, y_train.shape, y_test.shape

((404, 13), (102, 13), (404,), (102,))

2. Обучите стандартную регрессию, а также Ridge и  Lasso и параметрами по умолчанию и выведите их R2 на тестовой выборке

In [9]:
from sklearn.linear_model import LinearRegression

In [10]:
lr = LinearRegression()
lr.fit(X_train, y_train)
print (f'R2: {lr.score(X_test,y_test)}')

R2: 0.6687594935356314


In [11]:
from sklearn.linear_model import Ridge

In [12]:
ridge = Ridge()
ridge.fit(X_train,y_train)
print (f'R2: {ridge.score(X_test,y_test)}')

R2: 0.6662221670168518


In [13]:
from sklearn.linear_model import Lasso

In [14]:
lasso = Lasso()
lasso.fit(X_train,y_train)
print (f'R2: {lasso.score(X_test,y_test)}')

R2: 0.6671453631686304


3. Для Ridge и Lasso подберите коэффициент регуляризации(используйте GridSearchCV, RidgeCV, LassoCV) в пределах от $10^{-5}$ до $10^5$ (по степеням 10). Посчитайте R2 на тестовой выборке по лучшим моделям и сравните с предыдущими результатами. Напишите как изменился результат

In [15]:
from sklearn.model_selection import GridSearchCV

In [16]:
param = {'alpha':[1e-5, 1e-4, 1e-3, 1e-2, 1e-1, 1, 1e+2, 1e+3, 1e+4, 1e+5]}

search = GridSearchCV(Ridge(), param)
search.fit(X_train,y_train)
print(f'Best estimator: {search.best_estimator_}')

search = GridSearchCV(Lasso(), param)
search.fit(X_train,y_train)
print(f'Best estimator: {search.best_estimator_}')

Best estimator: Ridge(alpha=1e-05)
Best estimator: Lasso(alpha=1e-05)


In [17]:
from sklearn.linear_model import RidgeCV

In [18]:
ridge_cv = RidgeCV(alphas=[1e-5, 1e-4, 1e-3, 1e-2, 1e-1, 1, 1e+2, 1e+3, 1e+4, 1e+5])
ridge_cv.fit(X_train,y_train)
print (f'коэффициент регуляризации Ridge: {ridge_cv.alpha_}')

ridge = Ridge(alpha = ridge_cv.alpha_)
ridge.fit(X_train,y_train)
print (f'R2 тестовой выборки для alpha = {ridge_cv.alpha_}: {ridge.score(X_test,y_test)}')

коэффициент регуляризации Ridge: 0.01
R2 тестовой выборки для alpha = 0.01: 0.6687509462063148


In [19]:
from sklearn.linear_model import LassoCV

In [20]:
lasso_cv = LassoCV(alphas=[1e-5, 1e-4, 1e-3, 1e-2, 1e-1, 1, 1e+2, 1e+3, 1e+4, 1e+5])
lasso_cv.fit(X_train,y_train)
print (f'коэффициент регуляризации Lasso: {lasso_cv.alpha_}')

lasso = Lasso(alpha = lasso_cv.alpha_)
lasso.fit(X_train,y_train)
print (f'R2 тестовой выборки для alpha = {lasso_cv.alpha_}: {lasso.score(X_test,y_test)}')

коэффициент регуляризации Lasso: 1e-05
R2 тестовой выборки для alpha = 1e-05: 0.6687598638315153


*R2 стал как у  LinearRegression() без регуляризации*

4. Проведите масштабирование выборки(используйте Pipeline, StandardScaler, MinMaxScaler), посчитайте R2 и сравните с предыдущими результатами. Напишите как изменился результат

In [21]:
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import MinMaxScaler

In [22]:
pipe = make_pipeline(StandardScaler(), lr)
pipe.fit(X_train, y_train)
print (f'R2: {pipe.score(X_test,y_test)}')

R2: 0.668759493535632


In [23]:
pipe = make_pipeline(MinMaxScaler(), lr)
pipe.fit(X_train, y_train)
print (f'R2: {pipe.score(X_test,y_test)}')

R2: 0.6687594935356318


5. Подберите коэффициент регуляризации для Ridge и Lasso на масштабированных данных, посчитайте R2 и сравните с предыдущими результатами. Напишите как изменился результат

In [24]:
X_tr = StandardScaler().fit_transform(X)
X_train, X_test, y_train, y_test = train_test_split(X_tr, y, test_size=0.2, random_state=RANDOM_STATE)

In [25]:
ridge_cv.fit(X_train,y_train)
print (f'коэффициент регуляризации Ridge: {ridge_cv.alpha_}')

ridge = Ridge(alpha = ridge_cv.alpha_)
ridge.fit(X_train,y_train)
print (f'R2 тестовой выборки для alpha = {ridge_cv.alpha_}: {ridge.score(X_test,y_test)}')

коэффициент регуляризации Ridge: 1.0
R2 тестовой выборки для alpha = 1.0: 0.6684401592810275


In [26]:
lasso_cv.fit(X_train,y_train)
print (f'коэффициент регуляризации Lasso: {lasso_cv.alpha_}')

lasso = Lasso(alpha = lasso_cv.alpha_)
lasso.fit(X_train,y_train)
print (f'R2 тестовой выборки для alpha = {lasso_cv.alpha_}: {lasso.score(X_test,y_test)}')

коэффициент регуляризации Lasso: 1e-05
R2 тестовой выборки для alpha = 1e-05: 0.6687590128437058


6. Добавьте попарные произведения признаков и их квадраты (используйте PolynomialFeatures) на масштабированных признаках, посчитайте R2 и сравните с предыдущими результатами. Напишите как изменился результат

In [27]:
from sklearn.preprocessing import PolynomialFeatures

In [28]:
pipe = make_pipeline(PolynomialFeatures(), lr)
pipe.fit(X_train, y_train)
print (f'R2: {pipe.score(X_test,y_test)}')

R2: 0.805755340459623


7. Подберите наилучшую модель (используйте Pipeline, GridSearchSCV) подбирая тип регуляризации (L1,L2), коэффициент регуляризации, метод масштабирования и степень полинома в PolynomialFeatures. Выведите итоговые параметры и результат R2. Напишите как изменился R2 по сравнению с предыдущими экспериментами

In [29]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=RANDOM_STATE)

for L in [Ridge(), Lasso()]:
    search = GridSearchCV(L, param)
    search.fit(X_train,y_train)
    alpha_ = search.best_params_['alpha']
    print(f'L: {L} alpha: {alpha_} score: {search.score(X_train,y_train)}')

print(f'\nВыбираем Ridge(alpha = 1e-05)\n')
ridge = Ridge(alpha=1e-05)

for S in [StandardScaler(), MinMaxScaler()]:
    pipe = make_pipeline(S, ridge)
    pipe.fit(X_train, y_train)
    print (f'Scaler: {S} R2: {pipe.score(X_test,y_test)}')

print(f'\nВыбираем MinMaxScaler()\n')
scaler = MinMaxScaler()


for degree in [2, 3, 4, 5]:
    model = make_pipeline(scaler, PolynomialFeatures(degree), ridge)
    model.fit(X_train, y_train)
    print (f'Model: {model} R2: {model.score(X_test,y_test)}')

L: Ridge() alpha: 1e-05 score: 0.7508856358972874
L: Lasso() alpha: 1e-05 score: 0.7508856354745458

Выбираем Ridge(alpha = 1e-05)

Scaler: StandardScaler() R2: 0.668759490571006
Scaler: MinMaxScaler() R2: 0.668759628976253

Выбираем MinMaxScaler()

Model: Pipeline(steps=[('minmaxscaler', MinMaxScaler()),
                ('polynomialfeatures', PolynomialFeatures()),
                ('ridge', Ridge(alpha=1e-05))]) R2: 0.8064001291939079
Model: Pipeline(steps=[('minmaxscaler', MinMaxScaler()),
                ('polynomialfeatures', PolynomialFeatures(degree=3)),
                ('ridge', Ridge(alpha=1e-05))]) R2: 0.5103802100682169
Model: Pipeline(steps=[('minmaxscaler', MinMaxScaler()),
                ('polynomialfeatures', PolynomialFeatures(degree=4)),
                ('ridge', Ridge(alpha=1e-05))]) R2: 0.003803401008696472
Model: Pipeline(steps=[('minmaxscaler', MinMaxScaler()),
                ('polynomialfeatures', PolynomialFeatures(degree=5)),
                ('ridge', Ridge(alp

*Наилучший R2 показала модель с Ridge(alpha=1e-05), MinMaxScaler()) и PolynomialFeatures()
R2 c 0.6688 вырос до 0.8064*

http://archive.ics.uci.edu/ml/datasets/Adult

In [30]:
link = 'https://raw.githubusercontent.com/jbrownlee/Datasets/master/adult-all.csv'
data = pd.read_csv(link, header=None)

In [31]:
data.head()

0                 1       2          3   4                   5   \
0  39         State-gov   77516  Bachelors  13       Never-married   
1  50  Self-emp-not-inc   83311  Bachelors  13  Married-civ-spouse   
2  38           Private  215646    HS-grad   9            Divorced   
3  53           Private  234721       11th   7  Married-civ-spouse   
4  28           Private  338409  Bachelors  13  Married-civ-spouse   

                  6              7      8       9     10  11  12  \
0       Adm-clerical  Not-in-family  White    Male  2174   0  40   
1    Exec-managerial        Husband  White    Male     0   0  13   
2  Handlers-cleaners  Not-in-family  White    Male     0   0  40   
3  Handlers-cleaners        Husband  Black    Male     0   0  40   
4     Prof-specialty           Wife  Black  Female     0   0  40   

              13     14  
0  United-States  <=50K  
1  United-States  <=50K  
2  United-States  <=50K  
3  United-States  <=50K  
4           Cuba  <=50K

8. Разделите выборку на признаки и целевую переменную(колонка со зачениями {<=50K,>50K}). Замените целевую переменную на числовые значения.

In [32]:
X = data.iloc[:, :-1]
y = data.iloc[:, -1]

In [33]:
y.unique()

array(['<=50K', '>50K'], dtype=object)

In [34]:
y.replace(to_replace = ['<=50K', '>50K'], value = [0, 1], inplace = True)
y.unique()

array([0, 1])

9. Выясните, присутствуют ли в данных пропуски. Заполните их самыми частыми значениями (испольуйте SimpleImputer)

In [35]:
X.isnull().any().any()

False

10. Выберите колонки с числовыми и категориальными переменными.

In [36]:
numeric_features = [0, 2, 4, 10, 11, 12]
categorical_features = [1, 3, 5, 6, 7, 8, 9, 13]

11. Создайте пайплайн по обработке колонок(используйте OneHotEncoder,MinMaxScaler).

In [37]:
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import MinMaxScaler, OneHotEncoder

In [38]:
numeric_transformer = MinMaxScaler()
categorical_transformer = OneHotEncoder(handle_unknown="ignore")

preprocessor = ColumnTransformer(
    transformers=[
        ("num", numeric_transformer, numeric_features),
        ("cat", categorical_transformer, categorical_features),
    ]
)

12. Посчитайте метрики accuracy и f1_score на предсказании только самого частого класса в целевой переменной.

In [39]:
y.value_counts()

0    37155
1    11687
Name: 14, dtype: int64

In [40]:
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split

In [41]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=RANDOM_STATE)
X_train.shape, X_test.shape, y_train.shape, y_test.shape

((39073, 14), (9769, 14), (39073,), (9769,))

In [42]:
clf = Pipeline(steps=[("preprocessor", preprocessor), ("classifier", LogisticRegression(max_iter=1000))])
clf.fit(X_train, y_train)
y_pred = clf.predict(X_test)
print("model score: %.3f" % clf.score(X_test, y_test))

model score: 0.849


In [43]:
from sklearn.metrics import f1_score
from sklearn.metrics import accuracy_score

print(f'f1_score: {f1_score(y_test, y_pred, average=None, labels=[0, 1])[0]}')
print(f'accuracy_score: {accuracy_score(y_test, y_pred)}')

f1_score: 0.9035665384364607
accuracy_score: 0.8486027228989661


13. Посчитайте cross_val_score по алгоритмам LogisticRegression, SVC, LinearSVC по метрикам accuracy и f1_score.
Напишите удалось ли превзойти предыдущий результат.

In [44]:
from sklearn.model_selection import cross_val_score
cross_val_score(clf, X_train, y_train)
y_pred = clf.predict(X_test)
print(f'f1_score: {f1_score(y_test, y_pred, average=None, labels=[0, 1])[0]}')
print(f'accuracy_score: {accuracy_score(y_test, y_pred)}')

f1_score: 0.9035665384364607
accuracy_score: 0.8486027228989661


In [45]:
from sklearn.svm import SVC
clf = Pipeline(steps=[("preprocessor", preprocessor), ("classifier", SVC())])
clf.fit(X_train, y_train)
cross_val_score(clf, X_train, y_train)
y_pred = clf.predict(X_test)
print(f'f1_score: {f1_score(y_test, y_pred, average=None, labels=[0, 1])[0]}')
print(f'accuracy_score: {accuracy_score(y_test, y_pred)}')

f1_score: 0.8984870037501616
accuracy_score: 0.8392875422254069


In [46]:
from sklearn.svm import LinearSVC
clf = Pipeline(steps=[("preprocessor", preprocessor), ("classifier", LinearSVC())])
clf.fit(X_train, y_train)
cross_val_score(clf, X_train, y_train)
y_pred = clf.predict(X_test)
print(f'f1_score: {f1_score(y_test, y_pred, average=None, labels=[0, 1])[0]}')
print(f'accuracy_score: {accuracy_score(y_test, y_pred)}')

f1_score: 0.9050904830100247
accuracy_score: 0.8507523799774798


14. Можно заметить что в данных присутствуют значения '?', замените их самыми частыми значениями (испольуйте SimpleImputer)

In [47]:
X.replace("?", np.nan, inplace = True)
from sklearn.impute import SimpleImputer
imputer = SimpleImputer(missing_values=np.nan, strategy='most_frequent')
X_mf = imputer.fit_transform(X)

15. Посчитайте cross_val_score на новых данных. Напишите удалось ли улучшить результат.

In [48]:
from sklearn.svm import LinearSVC
clf = Pipeline(steps=[("preprocessor", preprocessor), ("classifier", LinearSVC())])
clf.fit(X_train, y_train)
cross_val_score(clf, X_train, y_train)
y_pred = clf.predict(X_test)
print(f'f1_score: {f1_score(y_test, y_pred, average=None, labels=[0, 1])[0]}')
print(f'accuracy_score: {accuracy_score(y_test, y_pred)}')

f1_score: 0.9050904830100247
accuracy_score: 0.8507523799774798


*Не удалось*

16. Посчитайте cross_val_score, если просто удалить значения '?'. Напишите как изменился результат

In [49]:
data = pd.read_csv(link, header=None)
df = data.replace("?", np.nan)
df.dropna(inplace=True)

In [50]:
X = df.iloc[:, :-1]
y = df.iloc[:, -1]

In [51]:
y.replace(to_replace = ['<=50K', '>50K'], value = [0, 1], inplace = True)

In [52]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=RANDOM_STATE)
X_train.shape, X_test.shape, y_train.shape, y_test.shape

((36177, 14), (9045, 14), (36177,), (9045,))

In [53]:
from sklearn.svm import LinearSVC
clf = Pipeline(steps=[("preprocessor", preprocessor), ("classifier", LinearSVC())])
clf.fit(X_train, y_train)
cross_val_score(clf, X_train, y_train)
y_pred = clf.predict(X_test)
print(f'f1_score: {f1_score(y_test, y_pred, average=None, labels=[0, 1])[0]}')
print(f'accuracy_score: {accuracy_score(y_test, y_pred)}')

f1_score: 0.904166368898735
accuracy_score: 0.8517412935323383


*Результат ухудшился*

 17. Посчитайте cross_val_score для RandomForestClassifier,GradientBoostingClassifier. Напишите как изменился результат и какой вывод можно из этого сделать.

In [54]:
from sklearn.ensemble import RandomForestClassifier
clf = Pipeline(steps=[("preprocessor", preprocessor), ("classifier", RandomForestClassifier())])
clf.fit(X_train, y_train)
cross_val_score(clf, X_train, y_train)
y_pred = clf.predict(X_test)
print(f'f1_score: {f1_score(y_test, y_pred, average=None, labels=[0, 1])[0]}')
print(f'accuracy_score: {accuracy_score(y_test, y_pred)}')

f1_score: 0.9043453112573205
accuracy_score: 0.8537313432835821


In [ ]:
from sklearn.ensemble import GradientBoostingClassifier
clf = Pipeline(steps=[("preprocessor", preprocessor), ("classifier", GradientBoostingClassifier())])
clf.fit(X_train, y_train)
cross_val_score(clf, X_train, y_train)
y_pred = clf.predict(X_test)
print(f'f1_score: {f1_score(y_test, y_pred, average=None, labels=[0, 1])[0]}')
print(f'accuracy_score: {accuracy_score(y_test, y_pred)}')

*Лучший результат у GradientBoostingClassifier*

18. Подберите наилучшую модель, подбирая методы обработки колонок - масштабирование признаков, кодирование признаков и заполнение пропусков. Параметры алгоритмов оставьте по умолчанию. Выведите итоговые параметры и результат accuracy и f1_score.

In [ ]:
print(f'масштабирование признаков - MinMaxScaler')
print(f'кодирование признаков - OneHotEncoder')
print(f'заполнение пропусков - dropna')
print(f'алгоритм - GradientBoostingClassifier')
print(f'итоговый f1_score: {f1_score(y_test, y_pred, average=None, labels=[0, 1])[0]}')
print(f'итоговый accuracy_score: {accuracy_score(y_test, y_pred)}')